### ISSUE TITLE
Complete set of Feature-Policy rules to consider

### ISSUE THREAD

**Issue by philwareham** [textpattern/server-config#38]
We have implemented some Feature-Policy rules - although the feature is in draft form and the spec and number of policies is in flux. Listed below are rules that we may want to consider as things develop. Please leave this issue open (but check off any rules we have implemented) as this will be on ongoing process. I have included links to documentation for each policy where available - feel free to add any missing links as and when you find them.

- [ ] accelerometer
- [ ] animations
- [ ] ambient-light-sensor
- [ ] [autoplay](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/autoplay)
- [ ] [camera](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/camera)
- [ ] [document-domain](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/document-domain)
- [ ] [encrypted-media](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/encrypted-media)
- [ ] [fullscreen](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/fullscreen)
- [ ] [geolocation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/geolocation)
- [ ] gyroscope
- [ ] lazyload
- [ ] magnetometer
- [ ] [microphone](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/microphone)
- [ ] [midi](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/midi)
- [ ] oversized-images
- [ ] [payment](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/payment)
- [ ] picture-in-picture
- [ ] speaker
- [ ] sync-script
- [ ] sync-xhr
- [ ] unoptimized-images
- [ ] unsized-media
- [ ] usb
- [ ] vertical-scroll
- [ ] [vr](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Feature-Policy/vr)

**Comment by petecooper** [comment]
Feature Policy has been renamed Permissions Policy:

https://scotthelme.co.uk/goodbye-feature-policy-and-hello-permissions-policy/

https://caniuse.com/permissions-policy (no browser support at the moment)
https://caniuse.com/feature-policy (~90% at time of writing)

**Comment by philwareham** [comment]
Thanks Pete - will investigate this next week.

**Comment by petecooper** [comment]
No action needed really, I'm doing Feature Policy to Permissions Policy duping as sites are migrated (given zero browser support it's hardly urgent), so we can run both in parallel for as long as we need to.

**Comment by philwareham** [comment]
@petecooper I think now is the time to remove Feature-Policy from the headers and just use Permissions-Policy. The inclusion of the older header is giving me console warnings in Chrome and affecting Lighthouse scores.

Since Chrome is the browser the most advanced terms of Feature-Policy/Permsisisons-Policy support which should IMO use that as the benchmark. Other browsers will gradually fall in line with the updated spec.

**Comment by petecooper** [comment]
Yep - agreed. I'll strip 'em out and upload this evening unless you get there first.

**Comment by petecooper** [comment]
Actually, I have a bit of time before a meeting, I'll start now.

**Comment by philwareham** [comment]
Cheers. I see we are also missing some of the previous rules such as `accelerometer` - any reason for that?

**Comment by petecooper** [comment]
Not intentionally, no - by all means add back in, I haven't taken stuff out knowingly!

**Comment by philwareham** [comment]
OK, if you can strip out feature policy I will then add the additional rules to Permission Policy, so we don't clash with commits. Thanks very much!

**Comment by petecooper** [comment]
OK, first pass of FP stuff done, I'll upload configs in the interim then do it again when you're done.

**Comment by philwareham** [comment]
@petecooper great, I have now updated all the Permissions-Policy rules as per our list above. Happy for this to move to the live servers when you have spare time. No rush.

**Comment by petecooper** [comment]
@philwareham - all uploaded just now, Nginx bounced.

**Comment by petecooper** [comment]
@philwareham - we (you) should also update the OP links to reflect PP directives, not FP directives.

**Comment by philwareham** [comment]
Thank you Pete! Have a good day.

**Comment by Bloke** [comment]
Continuing discussion from [here](https://forum.textpattern.com/viewtopic.php?pid=329874#p329874) then...

Seems we have various classes of policy but categorising them is not intuitive (as you've no doubt found!). Combined with that, we currently have two broad sets of policies:

1. For docs, RPC, Textile, .io, .net, .org:
    camera=()
    fullscreen=(self)
    geolocation=()
    gyroscope=()
    interest-cohort=()
    magnetometer=()
    microphone=()
    midi=()
    notifications=(self)
    payment=()
    push=(self)
    speaker=()
    sync-xhr=(self)
    vibrate=()
2. Additional policies for .com:
    accelerometer=()
    ambient-light-sensor=()
    autoplay=()
    battery=()
    display-capture=()
    picture-in-picture=()
    screen-wake-lock=()
    usb=()
    xr-spatial-tracking=()

BUT in the .com case there are three missing from the common list:
-   notifications
-  speaker
-  vibrate

So the 'cheap' way to do it is to just have 3 includes, one for each of the above sets and load them in according to how we need them per server.

Better, might be a classification system that we can use to lump newer things in as and when they come along but, as noted, that's fun especially given that some of the classes we need on one server and not another. I've had a go and come up with four potential groups:

### Peripherals - always shared (at the moment)
- Common
    camera
    microphone
    midi
- Omitted from .com
    speaker

### Mobile (out and about)
- Common
    geolocation
    gyroscope
    magnetometer
- Extras for .com
    accelerometer
    ambient-light-sensor
    battery
    screen-wake-lock
- Omitted from .com
    vibrate

### Display-related (notifications/updates)
- Common
    fullscreen(self)
    push(self)
    sync-xhr(self)
- Extras for .com
    display-capture
    picture-in-picture
- Omitted from .com
    notifications(self)

### Interfaces (to other software systems)
- Common
    interest-cohort
    payment
- Extras for .com
    autoplay
    xr-spatial-tracking
    usb

Bit of a mess. We could narrow that down to hardware/software split - that might be neater. I'm sure if we hammered this around a bit we could come up with a suitable taxonomy that allowed us to keep our hair/beards intact and manage this stuff in a sane way.

**Comment by petecooper** [comment]
>BUT in the .com case there are three missing from the common list:
>notifications
>speaker
>vibrate

They must've slipped through the net - not deliberately omitted afaik. Will fix.

**Comment by Bloke** [comment]
Ah, that makes it slightly easier as we only have two sets to maintain :)

**Comment by philwareham** [comment]
Please don’t action this until I’m back from holiday. The sets should be identical and also refect the list at the top of this thread (some policies never became final candidate so we’re removed, which I updated above). I need to double check if any more have been added or removed from spec in last 2 months before we can revise again.

**Comment by Bloke** [comment]
Cool, then that makes it simpler still :) One include. Job done.

**Comment by philwareham** [comment]
> They must've slipped through the net - not deliberately omitted afaik. Will fix.

Sorry, I've revert your commit - those rules were removed from the spec hence why they were not on .com site. I'll go through the other site configs and re one them too now. Cheers.

**Comment by philwareham** [comment]
FYI I've removed the `interest-cohort` policy from all but the forum domain, for reasons [I've already stated](https://forum.textpattern.com/viewtopic.php?pid=329899#p329899) in the forum.

**Comment by philwareham** [comment]
@petecooper hi Pete, I have just updated all the server config files to ensure they have latest (and consistent) set of rules. Can you deploy to live when you get a moment? Thanks!

**Comment by petecooper** [comment]
@philwareham - done. For future ref, please try to leave files in https://github.com/textpattern/server-config/tree/main/temp/_to%20check%20and%20purge as-is since I need to pick through for anything that's dropped by the wayside as things evolve. No worries in this case since it's permissions policy uniformity, but there's value in them being former production files – I take a few now and again and diff against the current files, and tweak as necessary (mostly it's a just a housekeeping exercise). Thanks!

**Comment by philwareham** [comment]
OK, will keep that in mind. Thanks Pete.
